Add support for fetching permission state from calling app launching the activity for system browser

quh4gko8 opened the pull request.

thestinger committed:
add exec-based spawning support

thestinger committed:
implement dynamic deny_new_usb toggle mode

thestinger committed:
core/config: enable doze and app standby

thestinger committed:
SettingsProvider: hide sensitive information while locked by default

thestinger committed:
SettingsProvider: set TEXT_SHOW_PASSWORD to false by default

thestinger committed:
increase default max password length to 64

thestinger committed:
add system property for disabling keyguard camera

Change-Id: I0b65cac3c3d2fc495b339c34add742bd698b107c

thestinger committed:
use permanent fingerprint lockout immediately

Signed-off-by: anupritaisno1 <www.anuprita804@gmail.com>

thestinger committed:
core/config: use round adaptive icon mask and parse round icons

Signed-off-by: anupritaisno1 <www.anuprita804@gmail.com>

thestinger committed:
SettingsLib: use Cloudflare DNS servers as hints

Cloudflare DNS has a better privacy policy than Google Public DNS while
still supporting DNS over TLS.

thestinger committed:
core/config: use Cloudflare DNS as the default fallback

Cloudflare DNS has a better privacy policy than Google Public DNS while
still supporting DNS over TLS.

MSe1969 committed:
SUPL: Don't send IMSI / Phone number to SUPL server

13: d93766fa56de

Change-Id: I5ccc4d61e52ac11ef33f44618d0e610089885b87

Squashed with:
Author: Daniel Micay <danielmicay@gmail.com>
Date:   Wed Mar 15 06:32:20 2023 -0400

    simplify removal of SUPL IMSI / phone number

    This is not required for SUPL to work and the comment about Google is
    unnecessary.

thestinger committed:
core/config: enable gesture navigation by default

thestinger committed:
core/config: enable multi user UI

thestinger committed:
core/config: raise maximum users to 32

thestinger committed:
SettingsProvider: disable bluetooth by default

thestinger committed:
make INTERNET into a special runtime permission

thestinger committed:
SystemUI: enable NFC quick settings tile

thestinger committed:
stop auto-granting location to system browsers

thestinger committed:
GrapheneOS boot logo mask

Updated by smdyv <88971341+smdyv@users.noreply.github.com>.

thestinger committed:
add parameter for avoiding full preload with exec

thestinger committed:
disable OpenGL preloading for exec spawning

thestinger committed:
disable resource preloading for exec spawning

thestinger committed:
disable class preloading for exec spawning

thestinger committed:
disable WebView reservation for exec spawning

thestinger committed:
disable JCA provider warm up for exec spawning

thestinger committed:
disable exec spawning when using debugging options

The debugging options are not yet supported probably, so disable exec
spawning when doing debugging.

thestinger committed:
pass through fullPreload to libcore

thestinger committed:
disable preloading classloaders for exec spawning

thestinger committed:
disable preloading HALs for exec spawning

rustybird committed:
core/config: disable safe volume everywhere, not just the US

Safe Volume has many false alarms. And if overridden by the user, at
every 20 hours of accumulated(!) audio playback it will intentionally
reset - even while the audio is playing. This can ruin the listening
experience, and ironically pose a safety hazard when e.g. the user is
operating a vehicle...

AOSP already opted out of Safe Volume in the overlays for Mobile Country
Code 310-316 (= US). Make this universal.

Signed-off-by: Rusty Bird <rustybird@net-c.com>

inthewaves committed:
SystemUI: have NfcTile get an NfcAdapter directly

This changes the NfcTile to not be dependent on an NfcManager giving the
NfcTile an NfcAdapter.

The problem with NfcTile is that the NfcService doesn't start until the
device is unlocked after a reboot, but NfcTile can be created and have
its state updated before the device is unlocked.

The state of NfcTile depends on an NfcAdapter. The tile gets an
NfcAdapter from a call to NfcAdapter#getDefaultAdapter, which gets an
adapter from an NfcManager via Context#getSystemService, and NfcManager
tries to get an adapter in its constructor via NfcAdapter#getNfcAdapter.

If this is done before unlock, NfcService isn't started, so the
constructor of NfcManager fails to get an adapter, opting to just store
null. This means that the NfcManager that's _cached_ by the NfcTile's
application context holds a null NfcAdapter, so subsequent calls in
NfcTile to get the NfcAdapter will keep returning null.

We can just have NfcTile get the NfcAdapter directly via
NfcAdapter#getNfcAdapter instead of relying on an NfcManager to call the
same method for us to get its default adapter. We just have to make sure
we use the application context for NfcAdapter#getNfcAdapter, as per the
doc comments for getNfcAdapter. This means that there's no longer an
NfcManager associated with the NfcTile's application context.

It doesn't look like NfcManager does anything special with the
NfcAdapter anyway. It seems to just be some middle man for NfcAdapters:
* The NfcAdapter field in NfcManager is final, and it doesn't do
anything else with it besides have a getter method for returning it. The
NfcManager seems to be a way to force third-party apps that want to get
an NfcAdapter to use the application context.
* NfcAdapter#getNfcAdapter manages the caching of NfcAdapters by
application context in a static HashMap. NfcManager doesn't manage
caching; it just calls NfcAdapter#getNfcAdapter.

kdrag0n committed:
Show USB icon in notification instead of generic system icon

It doesn't make sense to show a generic Android letter version icon for
USB.

Change-Id: I0441fc76fa8beab16675ac91e92e9b0490044dec

luk1337 committed:
Sharesheet: Display two rows of max ranked targets

This change makes sharesheet way more useful by increasing the amount of
visible ranked apps.

Change-Id: Ic092f1d1784259c9f3c0870eda1dd1ae8544c697

randomhydrosol committed:
pass through runtime flags for exec spawning and implement them in the child

Signed-off-by: anupritaisno1 <www.anuprita804@gmail.com>

thestinger committed:
add check for empty TTS engine name

thestinger committed:
core/config: set GrapheneOS launcher as notification listener

empratyush committed:
enable backup service for non-owner users

lzlrd committed:
Add a config to state whether a device supports increased touch sensitivity. (#2)

This is the partner commit to the addition of an option in Settings for
the same feature. This config can be enabled by an overlay for devices
that support increased touch sensitivity (otherwise known as "Glove
Mode") via the persist.vendor.touch_sensitivity_mode system property.

Signed-off-by: Diab Neiroukh <lazerl0rd@thezest.dev>

flawedworld committed:
CarrierConfig: Always show APN settings on CDMA carriers

thestinger committed:
core/config: revert to pre-12 power button behavior

kdrag0n committed:
SystemUI: Use new privacy indicators for location

Android has had location indicators for a while, but let's use the new
privacy indicator infrastructure for displaying them. This makes them
integrate better with the new camera and microphone indicators.

Change-Id: Ie457bb2dad17bddbf9dc3a818e3ec586023ce918

kdrag0n committed:
appop: Save discrete app op history for more permissions

Android 12's privacy dashboard shows permission usage timelines for
location, camera, and microphone. However, there's no reason to limit it
to those specific permissions; all the infrastructure is in place for
other permissions.

To enable the usage timeline for more permissions, keep discrete app op
history for all permission groups shown in the privacy dashboard. The
list of permission group -> app op mappings was obtained from
AppOpsManager.RUNTIME_AND_APPOP_PERMISSION_OPS with a few additional ops
from PrivacyItemController, and each op was resolved to its respective
enum ordinal from frameworks/proto_logging/stats/enums/app/enums.proto.

Change-Id: I6b1c476ea4c0edbc0b3fdf2e3e5cfcb11da77e33

thestinger committed:
SettingsProvider: enable backup support via seedvault

flawedworld committed:
enable secondary user logout support by default

Ported from 11, 12 moved the isLogoutEnabled boolean to ActiveAdmin.java

thestinger committed:
Revert "disable OpenGL preloading for exec spawning"

This reverts commit 5a8d91b5fac0a1ae597de359128e0706776ce3a7.

thestinger committed:
core/config: disable long press on power assistant setting

flawedworld committed:
SettingsProvider: Migration for legacy fingerprint keyguard toggle

thestinger committed:
SystemUI: change default quick tiles and quick tile order

thestinger committed:
core/config: [temporary] don't report statementservice crashes

There's a known issue in AOSP 12 caused by it creating overly large
intents for the work manager. It's not caused by anything done by end
users and they can't work around it so avoid reporting it to them since
they get a message each time it tries again.

flawedworld committed:
CarrierConfig: Do not mark dun APN types as read only

Previously I set this to null, but this can cause a NPE

thestinger committed:
stop enforcing camera sound for certain carriers

India, Japan and Korea have either industry standards or regulations for
phones sold within the country enforcing camera sounds. It's trivially
bypassed by taking out the SIM card, using video, using a headset or
turning off the volume. It doesn't make sense for us to enforce this.

kdrag0n committed:
Add config to exempt telephony-related app from location indicators

On some devices, there is a telephony-related app that frequently
requests location, so exempting it from privacy indicators is desirable.
For example, the Pixel 6 has an IMS service where this applies.

Change-Id: I5e99c89367bc3ffd31794736b0d66014fdc4faae

flawedworld committed:
CarrierConfig: Always show ICCID

thestinger committed:
use black wallpaper by default

kdrag0n committed:
colors: Switch to GrapheneOS blue color palette for Material You

Generated with Android 12 Extensions v9.0.0-test2 [1] using #1565C0
(light link accent color from GrapheneOS website) as a seed color,
with all other settings left at themelib [2] and colorkt [3] defaults.

[1] https://github.com/kdrag0n/android12-extensions/
[2] https://github.com/ProtonAOSP/android_external_themelib
[3] https://github.com/kdrag0n/colorkt

kdrag0n committed:
SystemUI: Enable production-ready feature flags
- Conversations: enabled by default on Pixel
- Charging ripple: enabled by default on Pixel
Taken from https://github.com/ProtonAOSP/android_vendor_proton/commit/a32d8c4c9e731e609b16b37da191669f5c607aad

thestinger committed:
improve PendingIntent security check compatibility

This switches to secure-by-default instead of crash-by-default for API
31 to work around apps which have updated to API 31 without specifying
either FLAG_MUTABLE or FLAG_IMMUTABLE for PendingIntents. If the app
ends up needing the FLAG_MUTABLE behavior, it may crash later, but it
should still be obvious why it happened.

There are many apps with outdated Play services client libraries lacking
support for Android 12 which are nonetheless targeting API 31 or higher
and will crash in certain situations. Google Play services will ask the
client library to request runtime permissions from the user on behalf of
it when it thinks that they're required for an operation that's
requested. The older client libraries will cause a crash in the app by
trying to create a PendingIntent with no FLAG_MUTABLE or FLAG_IMMUTABLE
specified. This is a much more common issue on GrapheneOS since Play
services is a regular user installed app with no special access or
privileges, and starts without any standard runtime permissions granted
to it.

Ported from 13: 94363af7c45a
Co-authored-by: Dmitry Muhomor <muhomor.dmitry@gmail.com>

x86pup committed:
Support forwarding notifications from other users

Commit history:

Before Android 13:
https://github.com/GrapheneOS/platform_frameworks_base/commits/30ab1eae4cd03270f0dce82a67a0122df7fb7c29

13: a40e627709a7

Signed-off-by: June <june@eridan.me>
Co-authored-by: inthewaves <inthewaves@pm.me>
Co-authored-by: June <june@eridan.me>
Co-authored-by: quh4gko8 <88831734+quh4gko8@users.noreply.github.com>
Co-authored-by: Pratyush <codelab@pratyush.dev>

muhomorr committed:
PackageInstaller: link "App info" screen from the uninstallation dialog

"App info" screen now has a "Disable" button, which can be used as an
alternative to uninstallation.

13: 309b0335cee6

thestinger committed:
exempt some system packages from status bar privacy indicator

x86pup committed:
Remove build fingerprint from screenshot EXIF

Signed-off-by: June <june@eridan.me>

muhomorr committed:
PackageInstaller: make sure install confirmation returns a result

13: c806278d60c2

flawedworld committed:
Warn the user if they are using a prototype Pixel device

Consumer Pixel devices have "MP1.0" as their hardware revision usually.
Devices stolen from Google will have "EVT", "PVT" or "DVT" set in
"ro.revision" by the bootloader. Additionally, check the secure boot
state prop set by the bootloader, "ro.boot.secure_boot", to ensure it is
set to "1", as typically pre-production devices won't have a blown
secure boot efuse, completely destroying any concept of verified boot.
Check both props, if either check fails, notify the user.

Squashed with: 778380bcc524954ed1b71dda8c762c458dfec087

Co-authored-by: Dmitry Muhomor <muhomor.dmitry@gmail.com>

empratyush committed:
SystemUI: add Battery share & Battery share quick tiles

muhomorr committed:
SettingsLib: allow Permission APEX to use ActionButtons and MainSwitch

muhomorr committed:
DeviceIdleJobsController: don't ignore whitelisted system apps

Only user app IDs were written to `mDeviceIdleWhitelistAppIds`, both initially and when
`PowerManager.ACTION_POWER_SAVE_WHITELIST_CHANGED` broadcast was received. All other places that
listen to that broadcast retrieve both user and system app IDs.

The only place where `mDeviceIdleWhitelistAppIds` array is checked is in `isWhitelistedLocked()`,
which is called only by `updateTaskStateLocked()` to check whether the app is on the device idle whitelist.

It's not clear why DeviceIdleJobsController ignores system apps.
File level comment doesn't mention the distinction between system and user apps:
"When device is dozing, set constraint for all jobs, except whitelisted apps, as not satisfied."
Comment for isWhitelistedLocked() does, however:
"Checks if the given job's scheduling app id exists in the device idle user whitelist."
However, that method is called for both system and user apps, and returns false for system apps
because only whitelist of user apps is checked. This leads to long delays for jobs that were
submitted by whitelisted system apps when device is in the Doze mode. No such delays happen with
whitelisted user apps.

Other places use a different naming for array of app IDs that includes only user apps,
eg `mDeviceIdleWhitelistUserAppIds`, not `mDeviceIdleWhitelistAppIds`.

I've looked through the Git history of DeviceIdleJobsController and JobSchedulerService, but didn't
find a reason for this behavior. Perhaps, system apps were exempted from device idle JobScheduler
restricitions in some other place previously, or this was a bug from the start.

Tested on an emulator with the Messaging app, which uses JobScheduler
during processing of incoming SMS:
1. Check that Messaging app is on system deviceidle whitelist:
```
$ dumpsys deviceidle whitelist | grep com.android.messaging
system-excidle,com.android.messaging,10090
system,com.android.messaging,10090
```
2. Simulate sending an SMS: it appears immediately
3. Simulate Doze mode: `$ dumpsys deviceidle force-idle`
4. Simulate sending an SMS again. Message doesn't appear, even if the Messaging app is open
5. Exit Doze mode: `$ dumpsys deviceidle unforce`. All pending messages appear immediately
6. Add Messaging app to the user whitelist:
```
$ dumpsys deviceidle whitelist +com.android.messaging
$ dumpsys deviceidle whitelist | grep com.android.messaging
system-excidle,com.android.messaging,10090
system,com.android.messaging,10090
user,com.android.messaging,10090
```
7. Simulate Doze mode again: `$ dumpsys deviceidle force-idle`
8. Simulate sending an SMS, note that it appears immediately this time

Also made a test system app to make sure that this issue isn't caused by low targetSdk of the
Messaging app (it targets SDK 24). Same issue with targetSdk 32 app.

In both cases, applying this patch fixes the issue.

muhomorr committed:
add helpers for system_server extensions

muhomorr committed:
Wi-Fi auto turn off

Co-authored-by: Pratyush <codelab@pratyush.dev>

muhomorr committed:
Bluetooth auto turn off

Co-authored-by: Pratyush <codelab@pratyush.dev>

muhomorr committed:
control access to Android/obb directory with a GosPackageState flag

muhomorr committed:
infrastructure for the Storage Scopes feature

13: b5ff11faa09e

muhomorr committed:
gosps: support for per-app GrapheneOS-specific persistent state

muhomorr committed:
improve compatibility of INTERNET special runtime permission

muhomorr committed:
infrastructure for spoofing self permission checks

quh4gko8 committed:
Set always-on and lockdown on by default on first vpn connection at system_server side

Don't require reboot or settings re-set for always on and lockdown to
take effect on first vpn connection. The requirement for reboot, re-set
at settings has been caused by a permission not granted or declared by
VpnDialogs, which caused the unexpected behavior. Prevent this by
checking the permission of local process instead.

muhomorr committed:
exec spawning: don't close the binder connection when the app crashes

When an unhandled exception occured, binder connections were closed with
IPCThreadState::stopProcess() before the invocation of java.lang.Thread#dispatchUncaughtException().
By default, that method tries to report the crash via ActivityManager#handleApplicationCrash(),
which always failed due to the closed binder connection.
This meant that the crash dialog was never shown and additional crash handling was skipped.

Zygote-based spawning never calls IPCThreadState::stopProcess().

muhomorr committed:
add an option to show the details of an application error to the user

Adds a "Show details" item to crash and ANR (app not responding) dialogs that takes the user to a
SystemUI activity which shows the error details and allows to copy them to the clipboard or to
export them via the standard sharing UI.

muhomorr committed:
add a wrapper for execveat(2)

Needed for exec spawning, to pass custom flags to the kernel.

muhomorr committed:
zygote: define runtime flags that depend on exec spawning

muhomorr committed:
exec spawning: support disabling hardened_malloc and extended VA space

thestinger committed:
SystemUI: fully enable location indicators by default

muhomorr committed:
pre-grant POST_NOTIFICATIONS to more system apps

thestinger committed:
core/config: decrease minimum color temperature for Night Light to 686K

muhomorr committed:
notify the user when sensors access is denied by OTHER_SENSORS perm

Notification is not shown if OTHER_SENSORS was explicitly denied by the user.

muhomorr committed:
PackageInstallerUI: an option to skip auto-grant of INTERNET permission

13: 6c6fa6316b29

empratyush committed:
core/config: add camera app QrTile activity as default

muhomorr committed:
extend original-package renaming to provider authorities for Vanadium

In the general case, ContentProvider authorities can't be renamed because they aren't required to be
based on the package name.

Chromium always forms ContentProvider authorities by prefixing them with its package name, and
relies on this invariant in code.

When its package is renamed by original-package handling code, statements like
String authority = context.getPackageName() + CONSTANT
become invalid.

Add a special-case for Vanadium to fix this.

muhomorr committed:
don't skip AOT compilation of apps that weren't recently used

muhomorr committed:
extend original-package renaming to static launcher shortcuts

muhomorr committed:
allow privileged installer to search for packages across all users

13: 358007e1b2ab

luk1337 committed:
core/config: Set config_clockFontFamily to empty string

This will match pre QPR1 clock font.

Change-Id: Idef5fb093d81f49e40f2fcf4df69f834804916ae

muhomorr committed:
don't allow updating system packages to the same versionCode

versionCode of many system packages, including privileged ones, is set to the current SDK version
and is thus not incremented during non-major OS upgrades.
This allowed to downgrade them to the older version that had the same versionCode.

x86pup committed:
SettingsProvider: Enable screen auto brightness by default

Signed-off-by: r3g_5z <june@girlboss.ceo>

muhomorr committed:
SystemUI: improve media projection dialog wording

Remove the mention of audio recording if the app doesn't have the permission for it.

13: 5f6c3a1d4589

muhomorr committed:
allow out-of-band package updates to use original-package system

13: 13fe843829ec

muhomorr committed:
disable package parser cache

This is needed for properly verifying updates of system packages.

muhomorr committed:
perform additional boot-time checks on system package updates

muhomorr committed:
require fs-verity when installing system package updates

muhomorr committed:
infrastructure for special runtime permissions

13: e6761e05a9ee

muhomorr committed:
add special runtime permission for other sensors

Ported from 12: 9d5a62ed573bc3c7be8b19445b372fed13533d0e

muhomorr committed:
extsettings: add a set of helper classes for defining system settings

x86pup committed:
add toggle for screenshot timestamp EXIF metadata

Signed-off-by: r3g_5z <june@girlboss.ceo>

thestinger committed:
core/config: disable using telephony-based auto time

thestinger committed:
reduce system clock drift warning

muhomorr committed:
add option to scramble PIN layout

Applies to device PIN, SIM PIN and SIM PUK input screens.

muhomorr committed:
add GNSS SUPL setting

13: 47eb7cbe1686

muhomorr committed:
allow privileged installers to maintain a global list of busy packages

Adds a global data structure that is accessible by privileged installers and allows them to avoid
installing the same package at the same time.

x86pup committed:
SystemUI: make screenshot sound use touch sounds setting

Signed-off-by: r3g_5z <june@girlboss.ceo>

muhomorr committed:
do not auto-grant Camera permission to the eUICC LPA UI app

Google's LPA that is shipped on GrapheneOS handles requesting the Camera permission at runtime,
which allows the user to give it a one-time grant.

13: 7bb7df2d2b13

x86pup committed:
CarrierConfig: allow toggling VoWiFi while roaming by default

Signed-off-by: r3g_5z <june@girlboss.ceo>

muhomorr committed:
don't crash apps that depend on missing Gservices provider

This allows apps that have minor dependency on GSF/GmsCore (EuiccSupportPixel, Google Camera etc)
to work without having any of GSF, GmsCore, Play Store installed.

Gservices provider is currently hosted by GSF, soon to be hosted by GmsCore.

muhomorr committed:
add hooks for modifying PackageManagerService behavior

13: 4e3e8c230773761183e97444020dbeb877131450

muhomorr committed:
integrate Google's EuiccSupportPixel package

Depends on commit: "don't crash apps that depend on missing Gservices provider"

muhomorr committed:
infrastructure for creating per-package seinfo overrides

13: 9fdbd47fd5bf

muhomorr committed:
require vendor camera extensions package to be a system app by default

This is a precaution in case ro.vendor.camera.extensions.package is set to an unexpected value.

muhomorr committed:
exec spawning: support runtime resource overlays

muhomorr committed:
do not allow disabling app visibility filtering

Needed for PackageManagerHooks.shouldFilterApplication() method.

muhomorr committed:
support denying special access to hardware accelerators by Google apps

muhomorr committed:
add GNSS PSDS setting

muhomorr committed:
SettingsLib: support replacing drawable in SelectorWithWidgetPreference

muhomorr committed:
AppBindArgs: infrastructure for passing extra args to app process init

muhomorr committed:
support redirecting ContentProvider calls to another provider

muhomorr committed:
infrastructure for the Contact Scopes feature

muhomorr committed:
don't run jobs that need connectivity in apps that lack INTERNET perm

Apps sometimes misbehave when INTERNET permission is revoked and a job that they scheduled with
a connectivity constraint is executed.

muhomorr committed:
SettingsProvider: add a helper class for migrating Global,Secure values

muhomorr committed:
add exec spawning sysprop to ExtSettings

muhomorr committed:
add connectivity checks setting and its migration from Settings.Global

muhomorr committed:
integrate Google's eUICC LPA package (EuiccGoogle)

- isolate EuiccGoogle from all non-system package via AppsFilter, which stops it from sending data
to Google through GmsCore. EuiccGoogle doesn't send data to Google directly
- keep EuiccGoogle disabled by default, but do not disable it after each reboot
- remove a misleading "device information will be sent to Google" message that appears before eSIM
download

thestinger committed:
gmscompat: add compatibility layer for unprivileged GMS

Originally authored by Danny Lin <danny@kdrag0n.dev> for inclusion in
GrapheneOS. It has since been substantially extended and partially
rewritten by Dmitry Muhomor <muhomor.dmitry@gmail.com> (primarily) and
Daniel Micay <danielmicay@gmail.com>.

Commit history:
(better viewed with git log --grep gmscompat $TAG_NAME)

March 2024 - June 2024:
https://github.com/GrapheneOS/platform_frameworks_base/commits/2024053100

November 2023 - March 2024:
https://github.com/GrapheneOS/platform_frameworks_base/commits/2024030300

August 2023 - November 2023:
https://github.com/GrapheneOS/platform_frameworks_base/commits/2023112900

May 2023 - August 2023:
https://github.com/GrapheneOS/platform_frameworks_base/commits/2023080800

February 2023 - May 2023:
https://github.com/GrapheneOS/platform_frameworks_base/commits/TQ2A.230505.002.2023052900

August 2022 - February 2023:
https://github.com/GrapheneOS/platform_frameworks_base/commits/TQ1A.230205.002.2023020600

June 2022 - August 2022:
https://github.com/GrapheneOS/platform_frameworks_base/commits/SP2A.220505.006.2022073000

October 2021 - June 2022:
https://github.com/GrapheneOS/platform_frameworks_base/commits/SP2A.220505.002.2022053100

July 2021 - October 2021:
https://github.com/GrapheneOS/platform_frameworks_base/commits/RQ3A.211001.001.2021100606

muhomorr committed:
enable LOCK_PIN_ENHANCED_PRIVACY setting by default

This setting disables animations in keyguard PIN input UI.

muhomorr committed:
SystemUI: specify config_sceenshotWorkProfileFilesApp

This resource is overridden with an overlay on stock OS to Google Files app.
If it's empty, SystemUI crashes after screenshot of a work profile app is taken.
Screenshot was still saved, crash happened when constructing post-screenshot UI.

"sceenshot" typo is upstream.

muhomorr committed:
add support for hooking LocationManager methods

muhomorr committed:
improve compat with apps that try to access network location provider

If there's no network location provider:
- redirect network location requests to one of available location providers
- stub out mock location management calls for network provider (TestProvider methods), since
redirecting them is infeasible

These changes apply only to third-party apps.

There's a relevant system feature (android.hardware.location.network) that is still listed as
available to improve app compatibility.
Removing it would not make this change redundant, because there are apps that try to use the network
location provider without declaring the need for this feature or checking its presence at runtime.

muhomorr committed:
add system package check methods to PackageUtils

muhomorr committed:
filter out non-system packages from location indicator exemptions

muhomorr committed:
filter out non-system packages from getLocationBypassPackages() result

muhomorr committed:
SystemUI: apply "Add users from lock screen" setting to guest users

quh4gko8 committed:
SettingsProviderHooks: Infrastructure for hooking settings state init

quh4gko8 committed:
Disable "Add users when locked" settings on every boot

quh4gko8 committed:
Disable instant apps feature on every boot

muhomorr committed:
add RemoteKeyProvisioning server override setting

muhomorr committed:
add hooks for modifying ActivityThread behavior

muhomorr committed:
add a test for the new per-app hardening features

Currently covers restriction of dynamic code execution from memory/storage and blocking of ptrace
access.

Testing is performed on three kinds of HardeningTestApp: targetSdk 27, current targetSdk and
preinstalled current targetSdk app (the latter is included on debuggable builds).

Preinstalled app is tested separately to make sure that restrictions can't be weakened for it,
regardless of GosPackageState values.

To run the test, execute `atest HardeningTest`.

muhomorr committed:
infrastructure for passing extra args to Zygote and WebViewZygote

muhomorr committed:
add base class for complex per-app switches

muhomorr committed:
infrastructure for passing SELinux flags to the kernel

muhomorr committed:
add per-app native debugging setting

Value of the previous global setting is used as a default value of the new per-app setting.

For system apps, native debugging is blocked unconditionally.

muhomorr committed:
add per-app dynamic code execution settings

- dynamic code execution from memory is blocked for all system apps except Vanadium
- dynamic code execution from storage is blocked for all system apps
- dynamic code execution in WebView (i.e. Vanadium JIT) is currently not restricted due to the lack
of support in Vanadium WebView.
- UI for these restrictions is not yet implemented

muhomorr committed:
add per-app setting for hardened_malloc and extended VA space

maade93791 committed:
Widevine provisioning setting

Signed-off-by: maade69 <70593890+maade69@users.noreply.github.com>

muhomorr committed:
ErrorDialogController: do not auto-dismiss crash dialogs

This is a workaround for a bug that auto-dismisses crash dialog for native crash almost immediately
after it is shown.

Crash dialogs are shown only for foreground apps by default, there's no need to auto dismiss them.

muhomorr committed:
support showing dexopt progress during boot

showDexoptProgressBootMessage() is called by libartservice in art repo during post-OTA dexopt.

muhomorr committed:
sscopes: temp: disable incomplete handling of READ_MEDIA_VISUAL_USER_SELECTED

Treat it same way other storage perms are treated for now.

quh4gko8 committed:
SystemUI: Allow customizing lock screen shortcuts via ThemePicker

muhomorr committed:
do not delete compiled code (OAT files) of hibernated apps

muhomorr committed:
change system_server label from "Android System" to "GrapheneOS"

This label is shown in header of notifications posted from system_server.

muhomorr committed:
disable OtaDexoptService

It's currently broken and shouldn't be used.

muhomorr committed:
add notification-based UI for the background dexopt job

- report job progress when it's running
- prompt the user to restart optimized apps after job completion
- schedule automatic restart of optimized apps via ActivityManager.killProcessesWhenImperceptible()
in case they aren't restarted manually

muhomorr committed:
infrastructure for per-package dex2oat compiler filter overrides

muhomorr committed:
always use speed dex2oat compiler filter for VPN service packages

This allows to skip restarting them after post-OTA asynchronous optimization, which some VPN service
hosts don't handle properly.

quh4gko8 committed:
Do not use filter calling uid parameter originally intended for instant apps

muhomorr committed:
add per-app memory tagging setting

muhomorr committed:
infrastructure for handling callbacks from logd to system_server

muhomorr committed:
add helper class for notifications that link AppSwitch UI

muhomorr committed:
add system API for getting app info by PID

muhomorr committed:
always stop app zygotes when app is being force-stopped

Some of per-app features (e.g. SELinux restrictions) require app process restart to be applied.

muhomorr committed:
fix rate limiting for non-app tombstones

All non-app tombstones were considered to have the same processName ("UNKNOWN"), which meant that
frequent tombstones from one system process rate-limited tombstones from all system processes.

muhomorr committed:
notify the user about notable system journal entries

Show notifications about the following events:
- kernel crash (from last_kmsg) (optional)
- file system check error
- system_server crash (Java or native)
- system app native crash (optional, except for memory tagging crashes)
- non-app process native crash (optional, except for memory tagging crashes)
- some of hardened_malloc app crashes (those that are caused by fatal_error() call in hardened_malloc
after it detects an app error)

These events are recorded in the system journal ("DropBox") already, this commit surfaces them to
the user.

muhomorr committed:
add helper class for launching non-standard Settings intents

muhomorr committed:
infrastructure for showing notifications about SELinux TSEC_FLAG denials

muhomorr commented on services/core/java/com/android/server/wm/WindowManagerHooks.java:
```diff
@@ -0,0 +1,71 @@
+        // Do not take into account of package visibility here.
+        long token = Binder.clearCallingIdentity();
+        try {
+            return permService.checkPermission(packageName, permission, userId) == PackageManager.PERMISSION_GRANTED;
```
It's better to return `checkPermission()` result directly, there might be more return values in the future.

quh4gko8 replied:
Okay, adjusting it accordingly

muhomorr commented on services/core/java/com/android/server/wm/ActivityClientController.java:
```diff
@@ -705,6 +705,22 @@ public String getLaunchedFromPackage(IBinder token) {
 
+    @Override
+    public boolean doesLaunchedFromPackageHavePermission(IBinder token, String permission) {
+        if (getLaunchedFromPackage(token) == null && !WindowManagerHooks.canAccessLaunchedFromPackagePermission()) {
```
Are apps that can access getLaunchedFromPackage() allowed to check for caller permission without this patch?

If no, then this patch would grant them additional access.

quh4gko8 replied:
Yes, via standard permission checks

quh4gko8 replied:
Nonetheless, it may be better to restrict it to intended callers, the system browsers for now

muhomorr commented on services/core/java/com/android/server/wm/WindowManagerHooks.java:
```diff
@@ -0,0 +1,71 @@
+        final int userId = UserHandle.getUserId(r.launchedFromUid);
+        var permService = LocalServices.getService(PermissionManagerServiceInternal.class);
+
+        // Do not take into account of package visibility here.
```
Wording error ("of" seems to be a leftover)

quh4gko8 replied:
Updated in the current revision

muhomorr commented on core/java/android/app/Activity.java:
```diff
@@ -6931,6 +6931,42 @@ public String getLaunchedFromPackage() {
+     * @hide
+     */
+    @UnsupportedAppUsage
+    @Nullable
```
boolean can't be nullable

muhomorr commented on core/java/android/app/Activity.java:
```diff
@@ -6931,6 +6931,42 @@ public String getLaunchedFromPackage() {
+     */
+    @UnsupportedAppUsage
+    @Nullable
+    public boolean doesLaunchedFromPackageHavePermission(String permission) {
```
`checkLaunchedFromPackagePermission` would be a better name I think, similar to `Context.checkCallingPermission()`

muhomorr committed:
show notification after app is blocked from accessing ptrace

"App info" screen for controlling ptrace access is linked from this notification.

muhomorr commented on core/java/android/app/Activity.java:
```diff
@@ -6931,6 +6931,19 @@ public String getLaunchedFromPackage() {
     }
 
+    /**
+     * Returns if the permission is granted on app that initially launched this activity.
```
Outdated comment

muhomorr commented on core/java/android/app/IActivityClientController.aidl:
```diff
@@ -191,4 +191,11 @@ interface IActivityClientController {
             in IBinder taskFragmentToken);
+
+    /**
+     * Return {@code true} if the calling package had a permission granted.
```
Outdated comment

muhomorr commented on core/java/android/app/Activity.java:
```diff
@@ -6931,6 +6931,19 @@ public String getLaunchedFromPackage() {
+     * Returns if the permission is granted on app that initially launched this activity.
+     *
+     * <p>In order to receive the launching app's package name permission state,
+     * the current activity is from a system browser, otherwise {@code SecurityException} is thrown.
```
Need to improve wording here

muhomorr committed:
aauto: define a restricted version of the MANAGE_USB permission

muhomorr committed:
define READ_DEVICE_SERIAL_NUMBER permission for Build.getSerial() access

Build.getSerial() is protected with the very broad READ_PRIVILEGED_PHONE_STATE permission.

Android Auto needs Build.getSerial() in some cases, but doesn't need most of the other privileges
that READ_PRIVILEGED_PHONE_STATE grants.

muhomorr committed:
add method for updating cached package permission state

Package permission state is updated automatically for all packages after events that might impact
it, e.g. after package install or uninstall, after storage volume mount, after OS update etc.

On GrapheneOS, per-package permission policy can be changed via GosPackageState packageFlags.
This new method is needed for updating the cached permission state manually after packageFlags
change.

muhomorr committed:
aauto: add definition of restricted BLUETOOTH_PRIVILEGED permission

muhomorr committed:
aauto: add definition of restricted WIFI_PRIVILEGED permission

muhomorr committed:
PackageHooks: infrastructure for per-package hooks in system_server

muhomorr committed:
aauto: define ASSOCIATE_COMPANION_DEVICES_RESTRICTED permission

muhomorr committed:
aauto: support allowing association with "automotive projection" devices

muhomorr committed:
aauto: support granting privileged permissions via PackageHooks

muhomorr committed:
aauto: support using aauto USB accessory activity as a default activity

muhomorr committed:
infrastructure for custom handling of known packages

muhomorr committed:
disable support for pre-approving PackageInstaller sessions

APK is not yet available at the time of pre-approval request, which is incompatible with the
"Allow Network permission" checkbox in the confirmation dialog.

See PackageInstaller.Session#requestUserPreapproval() for more info.

muhomorr committed:
add a method for sending BOOT_COMPLETED broadcast to a specific package

Android Auto requires this broadcast to be sent to complete its initialization.

muhomorr committed:
DropBoxManager: fix TreeSet API contract violations

Instances of FileList and EntryFile are stored in java.util.TreeSet.

TreeSet requires that compareTo() is consistent with equals().

muhomorr committed:
adjust definition of TelecomManager.getUserSelectedOutgoingPhoneAccount()

muhomorr committed:
aauto: add definition of restricted READ_PRIVILEGED_PHONE_STATE permission

muhomorr committed:
define package name of first-party app source

muhomorr committed:
define READ_LOGS_FULL permission for promptless logcat access

It's used by the LogViewer app.

muhomorr committed:
add APIs for launching LogViewer app

muhomorr committed:
add config for Pixel Camera Services

muhomorr committed:
add auto-reboot feature

Requires the corresponding changes to system/core and system/sepolicy.

muhomorr committed:
DropBox: skip spurious system_server_crash entries

muhomorr committed:
hide from Aurora Store packages that it's not allowed to update

muhomorr committed:
DropBox: skip fsck.f2fs results that don't contain errors

muhomorr committed:
use HTTPS X-Time header for time updates instead of NTP

Based on 57448ae65ee86fd218bd411947588b848658e5e8 and d17960e03d9b641dbed7ff330f77480a82d97faf

Co-authored-by: Renlord <me@renlord.com>
Co-authored-by: Daniel Micay <daniel.micay@grapheneos.org>

muhomorr committed:
extend auto-time setting to forced time updates

Original commit: c1d9385442ef0b620527f6e46e37bff35425b4d8

thestinger committed:
run full explicit GC in SystemUI and system_server after locking

ART has special support for System.gc() and Android already uses this to
purge memory after unlocking. It makes sense to purge as much as we can
when locking too.

muhomorr committed:
add UsbManager API for configuring USB port security state

See UsbExt definition and UsbPortSecurity setting (added by next commit) for more info.

muhomorr committed:
add setting for USB port security state

muhomorr committed:
SettingsLib: show more APK details on App info screen

muhomorr committed:
SettingsLib: add support for extending permission toggle UIs

This change is required by Settings app commits that add extra items to "All files access" and
"Install unknown apps" special access permission screens.

muhomorr committed:
allow PackageInstaller to use platform APIs

PackageInstaller is being made into an updatable app upstream, but it currently depends on several
platform APIs on GrapheneOS.

muhomorr committed:
SystemUI: require unlocking to use QS tiles by default

muhomorr committed:
SystemUI: allow to use some QS tiles while device is locked

quh4gko8 committed:
bugfix: properly acquire package state from original package

This resolves the NPE crash when updating an application under the new
package name that initially had the original package as package name.

muhomorr committed:
allow privileged installers to use getSharedLibraries(MATCH_ANY_USER)

muhomorr committed:
exec spawning: add workaround for late init of ART userfaultfd GC

Chromium browser and its derivatives setup a seccomp syscall filter in their isolated processes,
which blocks creation of new userfaultfds.

Since 14 QPR2, ART uses a new userfaultfd-based GC.

When zygote-based process spawning is used, userfaultfd GC is initialized before any of app's code
is executed, i.e. before Chromium's seccomp syscall filter is installed.

When exec spawning is used, userfaultfd GC initialization is delayed until first garbage collection.
Chromium's seccomp syscall filter is already installed at that point.

This leads to crashes of isolated Chromium processes (both browser and WebView), with the following
log messages:
 E cr_seccomp: ../../sandbox/linux/seccomp-bpf-helpers/sigsys_handlers.cc:**CRASHING**:seccomp-bpf failure in syscall
 E cr_seccomp: nr=0x11a arg1=0x80001 arg2=0xc arg3=0xffffffffffffffff arg4=0xc

As a workaround, perform early initialization of ART userfaultfd GC in isolated processes by calling
System.gc() before executing app's code. On Pixel 8, this increases startup latency by around 4 to
10 milliseconds.

quh4gko8 committed:
Support for getting the calling app's permission state for activities

This allows for apps that can already access the calling app identity that launchers their activity, along with other select system apps such as browser that are granted with such access, to check whenever a permission is granted, similar to the permission entry in activity section of AndroidManifest.xml.

muhomorr committed:
rename AppsFilterImpl.grantImplicitAccess method

This is done to have the build break in case grantImplicitAccess() starts to get used in more
places, which might weaken AppsFilter-based app isolation via PackageHooks.

muhomorr committed:
allow only first party app source and shell to update system packages

neelchauhan committed:
increase default max password length to 128

neelchauhan committed:
Disallow the camera shortcut toggle from the lockscreen

Closes https://github.com/GrapheneOS/os-issue-tracker/issues/2508

muhomorr committed:
disable support for adoptable storage

Adoptable storage is not available on any of the currently supported devices.

Handling of adoptable storage would complicate the duress password implementation.

muhomorr committed:
add a test for the duress password feature

muhomorr committed:
add a helper class for wiping eUICCs

muhomorr committed:
add duress password feature

muhomorr committed:
add lock_settings shell commands for testing the duress password feature

muhomorr committed:
duress password test: check that Weaver is wiped as part of duress wipe

muhomorr committed:
extend the upstream wipe-without-reboot implementation

muhomorr committed:
add lock_settings shell command option for printing executed Weaver ops

muhomorr committed:
extend excludedMethods list in AndroidPackageTest

muhomorr committed:
temp: unconditionally enable system crash notifications

muhomorr committed:
gmscompat: adjust to DynamiteLoader changes in GmsCore 24.22

APKs of Dynamite modules are now opened via ParcelFileDescriptor.open().

muhomorr committed:
change default UsbPortSecurity mode to CHARGING_ONLY_WHEN_LOCKED

muhomorr committed:
update lint baseline

muhomorr committed:
gmscompat: move enableNfc() to GmsModuleHooks

NFC code is getting modularized, it'll be an APEX in 15.

muhomorr committed:
Revert "SystemUI: have NfcTile get an NfcAdapter directly"

This reverts commit bf2289dd024540cdfbfe4986c6d4e7c42bf3051c.

muhomorr committed:
temp: disable boot-time dexopt progress UI

muhomorr committed:
Revert "Remove SystemUIDialog subclass from InternetDialog"

This reverts commit 7e259ee6c576164371babaef85db9a2f9be6f4ce.

Reverted change regressed the Internet quick settings tile in secondary users.

thestinger committed:
Revert "temp: unconditionally enable system crash notifications"

This reverts commit 6bddf9b3f1b6245cf9ee72a7bb5e9975d2cf8b9b.

muhomorr committed:
EuiccWipe: add missing null check for UiccSlotInfo

UiccSlotInfo can be null in some exceptional cases.

muhomorr committed:
fixup! extend the upstream wipe-without-reboot implementation

muhomorr committed:
Revert "temp: disable boot-time dexopt progress UI"

This reverts commit 95048fb5c4597b0e7c29b55c8363a46fe1cde4fa.

muhomorr committed:
DuressPasswordTest: remove device wipe from test preparation

Device wipe was added to make the test more reliable (e.g. it would fail if PIN/password is already
set), but it slows down the test too much, especially on user builds where adb is disabled by default.

muhomorr committed:
DuressPasswordTest: adjust to changed representation of WeaverRead.toString()

muhomorr committed:
DuressPasswordTest: fix test failure when device screen is initially turned off

muhomorr committed:
fix ApplicationInfo flag initialization being skipped in some cases

muhomorr committed:
fix crash of boot-time info dialog due to delayed font init

com.android.text.flags.use_optimized_boottime_font_loading flag is now enabled, which delays font
initialization.

muhomorr committed:
fixup! add option to scramble PIN layout

muhomorr committed:
Revert "[CDM][1/2] Modulize files"

This reverts commit 868276cbd31c2f725ffcede90396bb0937071485.

muhomorr committed:
Revert "Move association revoke logic to AssociationRevokeProcessor"

This reverts commit 5f07744a75223826f6e7dd13e8b292910ed32b97.

loadAssociationsFromDisk() change makes it call methods on mAssociationRevokeProcessor object in
some cases. mAssociationRevokeProcessor is null at boot-time.

This leads to NullPointerException chain-crashes during boot-time state restoration, which renders
the device unbootable.

muhomorr committed:
fixup! add setting for USB port security state

muhomorr committed:
add status callback to USB port security state API

muhomorr committed:
fixup! add setting for USB port security state

quh4gko8 committed:
bugfix: Use cast checking first before casting to PackageExt

There's another implementation for PackageExtIFace, namely PackageExtDefault and class casting exception is encountered in cases where this was the non-null instance acquired.

muhomorr committed:
move auto-reboot callback listener from SystemUI to system_server

This significantly simplifies the fw/base part of auto-reboot implementation and fixes the following
bug: auto-reboot timer was not started when the device was locked for the first time after setting
up a lock screen (e.g. on first boot). That bug was due to an assumption that "keyguard dismissed"
event is always produced at least once before device is locked in an AFU state, which is not the
case when the device doesn't have a lockscreen when it transitions from BFU to AFU.

This commit fixes that bug, since onKeyguardShowingStateChanged() callback is invoked at boot even
when the device doesn't have a lockscreen.

muhomorr committed:
don't toggle USB-C port after first unlock if it was connected BFU

This is a regression that has appeared due to changed ordering between USB connection notification
and "keyguard shown" event.

muhomorr committed:
fixup! add setting for USB port security state

This is an overly verbose logging statement, that event is logged elsewhere already.
